File: app/src/main/java/name/monwf/customiuizer/prefs/DropDownPreferenceEx.java
package name.monwf.customiuizer.prefs;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import androidx.preference.DropDownPreference;
import androidx.preference.PreferenceViewHolder;

import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import name.monwf.customiuizer.R;
import name.monwf.customiuizer.utils.Helpers;

public class DropDownPreferenceEx extends DropDownPreference implements PreferenceState {

	private CharSequence sValue;
	private final Resources res = getContext().getResources();
	private final int primary = res.getColor(R.color.preference_primary_text, getContext().getTheme());
	private final int secondary = res.getColor(R.color.preference_secondary_text, getContext().getTheme());
	private final int childpadding = res.getDimensionPixelSize(R.dimen.preference_item_child_padding);

	private final int indentLevel;
	private final boolean dynamic;
	private boolean newmod = false;
    private boolean highlight = false;
	private boolean unsupported = false;
	private final boolean valueAsSummary;

	public DropDownPreferenceEx(Context context, AttributeSet attrs) {
		super(context, attrs);
		final TypedArray xmlAttrs = context.obtainStyledAttributes(attrs, R.styleable.ListPreferenceEx);
		indentLevel = xmlAttrs.getInt(R.styleable.ListPreferenceEx_indentLevel, 0);
		dynamic = xmlAttrs.getBoolean(R.styleable.ListPreferenceEx_dynamic, false);
		valueAsSummary = xmlAttrs.getBoolean(R.styleable.ListPreferenceEx_valueAsSummary, false);
		xmlAttrs.recycle();
		setIconSpaceReserved(false);
	}

	@Override
	public void setValue(String value) {
		super.setValue(value);
		int index = findIndexOfValue(value);
		if (index < 0 || index > getEntries().length - 1) return;
		sValue = getEntries()[index];
	}

	public void setUnsupported(boolean value) {
		unsupported = value;
		setEnabled(!value);
	}

	public void getView(View finalView) {
		TextView title = finalView.findViewById(android.R.id.title);
		TextView summary = finalView.findViewById(android.R.id.summary);
		TextView valSummary = finalView.findViewById(android.R.id.hint);

		summary.setVisibility(valueAsSummary || getSummary() == null || getSummary().equals("") ? View.GONE : View.VISIBLE);
		valSummary.setVisibility(valueAsSummary ? View.VISIBLE : View.GONE);
		valSummary.setText(valueAsSummary ? sValue : "");
		if (valueAsSummary) valSummary.setTextColor(Helpers.isNightMode(getContext()) ? secondary : primary);
		title.setTextColor(isEnabled() ? primary : secondary);
		title.setText(getTitle() + (unsupported ? " ⨯" : (dynamic ? " ⟲" : "")));
		if (newmod) Helpers.applyNewMod(title);
		if (highlight) {
			Helpers.applySearchItemHighlight(finalView);
		}
		int hrzPadding = (indentLevel + 1) * childpadding;
		finalView.setPadding(hrzPadding, 0, childpadding, 0);
	}

	@Override
	public void onBindViewHolder(PreferenceViewHolder view) {
		super.onBindViewHolder(view);
		TextView title = (TextView) view.findViewById(android.R.id.title);
		title.setMaxLines(3);

		TextView summary = (TextView) view.findViewById(android.R.id.summary);
		summary.setTextColor(secondary);

		TextView valSummary = view.itemView.findViewById(android.R.id.hint);
		if (valSummary == null) {
			valSummary = new TextView(getContext());
			valSummary.setTextSize(TypedValue.COMPLEX_UNIT_PX, summary.getTextSize());
			valSummary.setTextColor(summary.getCurrentTextColor());
			valSummary.setPadding(summary.getPaddingLeft(), summary.getPaddingTop(), res.getDimensionPixelSize(R.dimen.preference_summary_padding_right), summary.getPaddingBottom());
			valSummary.setId(android.R.id.hint);
			((ViewGroup) view.itemView).addView(valSummary, 2);
		}

		getView(view.itemView);
	}

	@Override
	public void markAsNew() {
		newmod = true;
	}

	@Override
	public void applyHighlight() {
		highlight = true;
	}
}

File: app/src/main/java/tv/withaibuild/customiuizer/prefs/DropDownPreferenceEx.kt
package tv.withaibuild.customiuizer.prefs

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.preference.DropDownPreference
import androidx.preference.PreferenceViewHolder
import tv.withaibuild.customiuizer.R
import tv.withaibuild.customiuizer.utils.Helpers

class DropDownPreferenceEx(context: Context, attrs: AttributeSet?) : DropDownPreference(context, attrs), PreferenceState {

    private var sValue: CharSequence? = null
    private val res = context.resources
    private val primary = res.getColor(R.color.preference_primary_text, context.theme)
    private val secondary = res.getColor(R.color.preference_secondary_text, context.theme)
    private val childPadding = res.getDimensionPixelSize(R.dimen.preference_item_child_padding)

    private val indentLevel: Int
    private val dynamic: Boolean
    private var newmod = false
    private var highlight = false
    private var unsupported = false
    private val valueAsSummary: Boolean

    init {
        val xmlAttrs: TypedArray = context.obtainStyledAttributes(attrs, R.styleable.ListPreferenceEx)
        indentLevel = xmlAttrs.getInt(R.styleable.ListPreferenceEx_indentLevel, 0)
        dynamic = xmlAttrs.getBoolean(R.styleable.ListPreferenceEx_dynamic, false)
        valueAsSummary = xmlAttrs.getBoolean(R.styleable.ListPreferenceEx_valueAsSummary, false)
        xmlAttrs.recycle()
        isIconSpaceReserved = false
    }

    override fun setValue(value: String?) {
        super.setValue(value)
        val index = findIndexOfValue(value)
        val entries = entries
        if (entries != null && index >= 0 && index < entries.size) {
            sValue = entries[index]
        }
    }

    fun setUnsupported(value: Boolean) {
        unsupported = value
        isEnabled = !value
    }

    fun getView(finalView: View) {
        val title = finalView.findViewById<TextView>(android.R.id.title)
        val summary = finalView.findViewById<TextView>(android.R.id.summary)
        val valSummary = finalView.findViewById<TextView>(android.R.id.hint)

        summary?.visibility = if (valueAsSummary || summary == null || summary.text.isNullOrEmpty()) View.GONE else View.VISIBLE
        valSummary?.visibility = if (valueAsSummary) View.VISIBLE else View.GONE
        valSummary?.text = if (valueAsSummary) sValue else ""
        if (valueAsSummary) valSummary?.setTextColor(if (Helpers.isNightMode(context)) secondary else primary)
        title?.setTextColor(if (isEnabled) primary else secondary)
        title?.text = (title?.text?.toString() ?: "") + if (unsupported) " ⨯" else if (dynamic) " ⟲" else ""
        if (newmod) title?.let { Helpers.applyNewMod(it) }
        if (highlight) Helpers.applySearchItemHighlight(finalView)
        val hrzPadding = (indentLevel + 1) * childPadding
        finalView.setPadding(hrzPadding, 0, childPadding, 0)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        val title = holder.findViewById(android.R.id.title) as? TextView
        title?.maxLines = 3

        val summary = holder.findViewById(android.R.id.summary) as? TextView
        summary?.setTextColor(secondary)

        var valSummary = holder.itemView.findViewById<TextView>(android.R.id.hint)
        if (valSummary == null) {
            valSummary = TextView(context).apply {
                setTextSize(TypedValue.COMPLEX_UNIT_PX, summary?.textSize ?: textSize)
                setTextColor(summary?.currentTextColor ?: currentTextColor)
                setPadding(summary?.paddingLeft ?: 0, summary?.paddingTop ?: 0, res.getDimensionPixelSize(R.dimen.preference_summary_padding_right), summary?.paddingBottom ?: 0)
                id = android.R.id.hint
            }
            (holder.itemView as? ViewGroup)?.addView(valSummary, 2)
        }

        getView(holder.itemView)
    }

    override fun markAsNew() {
        newmod = true
    }

    override fun applyHighlight() {
        highlight = true
    }
}
